frontend/src/components/editor/app-container.tsx

```diff
@@ -20,6 +27,52 @@ export const AppContainer: React.FC<PropsWithChildren<Props>> = ({
   isRunning,
   children,
 }) => {
+  const errors = useCellErrors();
+
+  // Dynamically update favicon for run feedback
+  let favicon: HTMLLinkElement | null =
+    document.querySelector("link[rel~='icon']");
+
+  if (!favicon) {
+    favicon = document.createElement("link");
+    favicon.rel = "icon";
+    document.getElementsByTagName("head")[0].append(favicon);
+  }
+
+  const favicons: Record<string, string> = useMemo(
+    () => ({
+      idle: "./favicon.ico",
+      success: "./circle-check.ico",
+      running: "./circle-play.ico",
+      error: "./circle-x.ico",
+    }),
+    [],
+  );
+
+  const resetFaviconIfComplete = useCallback(() => {
+    if (!isRunning && document.visibilityState == "visible") {
+      setTimeout(() => {
+        favicon.href = favicons.idle;
+      }, 3000);
+    }
+  }, [favicon, favicons.idle, isRunning]);
+
+  useEffect(() => {
+    if (isRunning) {
+      favicon.href = favicons.running;
+      return;
+    }
+    favicon.href = errors.length > 0 ? favicons.error : favicons.success;
+    resetFaviconIfComplete();
+    return () => {
+      favicon.href = favicons.idle;
+    };
+  }, [isRunning, errors, favicon, favicons, resetFaviconIfComplete]);
+
+  useEventListener(document, "visibilitychange", (_) => {
```

Comment:
can you document how you want this to work?

i think i mentioned we could use `document.visibilityState/visibilitychange`, but that was so when the editor is active, we arent flickering the favicon and only change it when the editor is out of focus. that was only a suggestion, but up-to-you what you feel feels right.

Reply:
also - im not sure this actually solves the ticket yet. i think what might happen with the above is:
1. goes to running
2. shows the success
3. then clears after 3 seconds

If you miss that 3 second window, you don't really get the feedback. Maybe the logic could be:

1. goes to running
2. shows the success
3. then clears after 3 seconds ONLY if the document is focused, otherwise keep it there
4. clear when the user comes back into focus (could be after 3 seconds or right away)

also worth documenting this logic in a comment on the `DynamicFavicon` component

Reply:
Sure, I'll add some comments in the file.

You're right, the favicon is reset after 3 seconds if the page is visible, but remains if it is not. I agree that it should persist if the document is not in focus; I'll add that in.